module/zfs/vdev.c

```diff
@@ -3670,6 +3671,116 @@ vdev_deadman(vdev_t *vd)
 	}
 }
 
+/*
+ * Implements the per-vdev portion of manual TRIM. The function passes over
+ * all metaslabs on this vdev and performs a metaslab_trim_all on them. It's
+ * also responsible for rate-control if spa_man_trim_rate is non-zero.
+ *
+ * If fulltrim is set, metaslabs without spacemaps are also trimmed.
```

Comment:
Just FYI, this is a ZoL change. The version from Nexenta always trims everything, for two very simple reasons:
1) We wanted to allow this to be used for pre-initializing empty space.
2) The cost of doing a trim/unmap on already trimmed sectors is very low (the device knows what blocks have been touched).